oembed.js

```diff
@@ -1,156 +1,153 @@
-var xml2js = require('xml2js');
-var async = require('async');
-var cheerio = require('cheerio');
+const fetch = require('node-fetch');
+const urls = require('url');
+const xml2js = require('xml2js');
+const async = require('async');
+const cheerio = require('cheerio');
+const util = require('util');
 
-var forceXml = false;
+let forceXml = false;
 
 module.exports = oembed;
 
 // The _canonical option is used internally to prevent
 // infinite recursion when retrying with a canonical URL.
 // Don't worry about it in your code.
 
-function oembed(url, options, endpoint, mainCallback, _canonical) {
-  var oUrl;
-  var result;
-  return async.series({
-    discover: function(callback) {
-      // if we're being told the end point, use it
-      if (endpoint)
-      {
-        if (!options) {
-          options = {};
-        }
+async function oembed(url, options, endpoint, callback, _canonical) {
+
+  try {
+    let result;
+    const { canonical, oUrl } = await discover();
+    if (canonical) {
+      return oembed(canonical, options, endpoint, callback, true);
+    }
+    return callback(null, await retrieve());
+  } catch (e) {
+    return callback(e);
+  }
 
-        oUrl = endpoint;
-        options.url = url;
-        return callback(null);
+  async function discover() {
+    // if we're being told the end point, use it
+    if (endpoint)
+    {
+      if (!options) {
+        options = {};
       }
 
-      // otherwise discover it
-      return request(url, {
-          headers: Object.assign({
-            'User-Agent': 'oembetter'
-          }, options.headers || {})
-        }, function(err, response, body) {
-        if (err) {
-          return callback(err);
-        }
-        var $ = cheerio.load(body);
-
-        // <link rel="alternate" type="application/json+oembed" href="http://www.youtube.com/oembed?format=json&amp;url=http%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dzsl_auoGuy4" title="An Engineer&#39;s Guide to Cats 2.0 - The Sequel">
-
-        // Allow for all the dumb stuff we've seen.
-        // (Only application/json+oembed and
-        // text/xmloembed are in the standard.)
-
-        var ideas = [
-          'link[type="application/json+oembed"]',
-          'link[type="text/json+oembed"]',
-          'link[type="application/xml+oembed"]',
-          'link[type="text/xml+oembed"]'
-        ];
-
-        var i;
-        for (i = 0; (i < ideas.length); i++) {
-          oUrl = $(ideas[i]).attr('href');
-          if (oUrl) {
-            oUrl = urls.resolve(response.request.href, oUrl);
-            if (url.match(/^https:/) && oUrl.match(/^http:/)) {
-              // Fix for YouTube's bug 12/15/20: issuing HTTP discovery URLs
-              // but flunking them with a 403 when they arrive
-              if (oUrl.match(/youtube/) && oUrl.match(/^http:/)) {
-                oUrl = oUrl.replace(/^http:/, 'https:');
-              }
-            }
-            break;
-          }
-        }
+      oUrl = endpoint;
```

Comment:
`oUrl` not explicitly declared. It might be worth adding eslint or running this through a linter following these structural changes.